### ISSUE TITLE
Trouble getting BM70 to be recognized

### ISSUE THREAD

**Issue by mumrah** [hessu/aprs-specs#1]
Not sure if this is the correct forum for this kind of question, please feel free to close this issue if its not.

I'm working on configuring a Microchip BM70 bluetooth module in the correct way to be discovered by the aprs.fi iOS app (and any other applications implementing the BLE spec defined in this repo).

As far as I can tell, I have the correct service and characteristic UUIDs:

![IMG_0789](https://user-images.githubusercontent.com/55116/94608807-f039ad00-026b-11eb-815b-8485c54d945e.PNG)

I noticed that the tnc3 firmware sets a few other pieces of info. I'm wondering if maybe this "magic" string is needed by aprs.fi https://github.com/mobilinkd/tnc3-firmware/blob/master/TNC/bm78.cpp#L308. I also notice that the UUIDs set in this firmware are different from those defined in the spec. 🤔 

Any guidance would be greatly appreciated!

**Comment by hessu** [comment]
These are the only magic UUIDs that I have in the aprs.fi iOS app source code, and they do match the UUIDs in the spec. The initial scan only looks for the service UUID so only that needs to be right for the device to appear in the discovery results list.

/* BLE UUIDs from ge0rg, implemented in mobilinkd too */
#define KTS_SERVICE_UUID @"00000001-ba2a-46c9-ae49-01b0961f68bb"
// Notify, Read
#define KTS_RX_CHAR_UUID @"00000003-ba2a-46c9-ae49-01b0961f68bb"
// Write/Response
#define KTS_TX_CHAR_UUID @"00000002-ba2a-46c9-ae49-01b0961f68bb"

Looking at the tnc3 firmware, the UUIDs seem to be in bm78_eeprom.cpp (in reverse byte order, obviously...!) and whatever appears in bm78.cpp is something else than the actual UUIDs.

**Comment by mumrah** [comment]
@hessu, I'm wondering if somehow my module isn't getting advertised properly. Can you share any code snippets from your app or other sources pertaining to BLE discovery? Maybe I'll find something that I'm doing wrong.

It's strange since I can discover and connect to the device from other BLE apps. I'll poke around in the tnc3's bm78 eeprom and see if i find any clues.

**Comment by hessu** [comment]
My code isn't substantially different from the TNC3 config app: https://github.com/mobilinkd/iosTncConfig/blob/master/Mobilinkd%20TNC%20Config/BLECentralViewController.swift#L172

I'm simply kicking off a discovery for the service UUID, and display all devices that are found.
    [_centralManager scanForPeripheralsWithServices:@[[CBUUID UUIDWithString:KTS_SERVICE_UUID]] options:@{ CBCentralManagerScanOptionAllowDuplicatesKey : @YES }];

You could try the Mobilinkd TNC3 config app, it seems like it should find and display all other TNCs which use the same UUID, regardless of manufacturer.

One thing to check: Does the aprs.fi app have access to Bluetooth in iOS Settings -> Privacy -> Bluetooth -> aprs.fi ? Noticed the app does not show an error message if it doesn't have permission. I should add a warning in the discovery screen.

**Comment by mumrah** [comment]
I was missing the service UUID from the advertisement packet. Now aprs.fi and iosTncConfig apps can see the device. However, once I connect from aprs.fi, I immediately see a disconnect with the reason: "0x13 Remote User Terminated Connection". 

![image](https://user-images.githubusercontent.com/55116/94696910-908fe000-0305-11eb-8760-5cf58adefa3e.png)

Here's the portion of the GATT table from the the Microchip manual test tool.

**Comment by mobilinkd** [comment]
Do you get the same disconnect from the iosTncConfig?

This is what nRF Connect on Android shows for the BLE service of a TNC3.

![nRFConnect_sm](https://user-images.githubusercontent.com/4713996/94756419-1be89000-035d-11eb-9c3d-6905156f6fbe.png)

**Comment by mumrah** [comment]
🎉 After a lot of trial and error, I was able to get the device recognized! 🎉

I was missing a few things. Mainly, I was missing the 0x2902 client descriptor (CCCD) on the RX characteristic. This is apparently needed for a few different client related things (including notifications). This was a total mystery to me since I am a BLE noob. It's apparently documented here as a well-known string https://www.bluetooth.com/specifications/gatt/descriptors/ 🤷 

Beyond that, I also needed to add the "read" and "write" properties to the RX characteristic. Otherwise when then apps would try to set the notify flag on the CCCD it would fail. By default I had value of the 0x2902 CCCD as 0x0000, and when a client sets notifications, it gets updated to 0x0100 (or similar). Without the write flag, the mobilinkd iOS config app, aprs.fi, and other BLE apps would fail to set the notification. I suspect this is an implementation detail of the BM70 since it seems like the BM78 works without this property on the RX characteristic.

I'll update this thread with more details and screenshots of the BM70 config later today. Thanks so much for your help @hessu and @mobilinkd! Now to write some Arduino code for talking to the BM70 😁
